src/structure.pyx

```diff
@@ -416,6 +419,102 @@ cdef class Structure:
+                   options = Structure.defaultStructureArrayOptions,
+                   classifier = None, *,
+                   selection = None):
+    """
+    Create array of structures from PDB file.
+
+    Split PDB file into several structures by either by treating
+    chains separately and/or by treating each MODEL as a separate
+    structure and/or grouping chains.
+
+    Args:
+        fileName (str): The PDB file.
+        options (dict): Specification for how to read the PDB-file
+            (see :py:attr:`.Structure.defaultStructureArrayOptions` for
+            options and default value).
+        classifier: :py:class:`.Classifier` to assign atoms radii, default is used
+            if none specified.
+        selection (str): chain group selection.
+
+    Returns:
+        list: An array of :py:class:`.Structure`
+
+    Raises:
+        AssertionError: if `fileName` is None
+        AssertionError: if an option value is not recognized
+        AssertionError: if neither of the options `'separate-chains'`
+            and `'separate-models'` are specified.
+        IOError: if can't open file
+        Exception: if there are problems parsing the input
+    """
+
+    assert fileName is not None
+    # we need to have at least one of these
+    assert(('separate-chains' in options and options['separate-chains'] is True)
+           or ('separate-models' in options and options['separate-models'] is True)
+           or (not selection is None))
+    structure_options = Structure._get_structure_options(options)
+    cdef FILE *input
+    input = fopen(fileName,'rb')
+    if input is NULL:
+        raise IOError("File '%s' could not be opened." % fileName)
+    
+
+    verbosity = getVerbosity()
+
+    if classifier is not None:
+        setVerbosity(silent)
+    
+    cdef int n
+    cdef freesasa_structure** sArray
+    if options.get('separate-chains', False) or options.get('separate-models', False):
+        sArray = freesasa_structure_array(input,&n,NULL,structure_options)
+    else:
+        sArray = <freesasa_structure **> malloc(sizeof(freesasa_structure *))
+        sArray[0] = freesasa_structure_from_pdb(input, NULL, structure_options)
+        n = 1
+    fclose(input)
+
+    cdef freesasa_structure* group
+    if selection is not None:
+        # Check input
+        if options.get('separate-chains', False):
+            raise ValueError("'chain_group' and 'separate-chains' cannot be set simultaneously.")
+        allowed_chars = set(string.ascii_lowercase + string.ascii_uppercase + '+')
+        if not all([character in allowed_chars for character in selection]):
+            raise ValueError("'{}' is not a valid chain-groups selection.".format(selection))
+        
+        # Add groups from each model
+        chain_groups = selection.split('+')
+        n_groups = len(chain_groups)
+        n_total = n * (1 + n_groups)
+        sArray = <freesasa_structure**> realloc(sArray, n_total * sizeof(freesasa_structure*))
```

Comment:
It's unlikely of course, but `realloc` can also return NULL. It's probably a bit overkill that the C code handles these cases, but since it does, we might as well do it here too.

Reply:
Added check [here](https://github.com/freesasa/freesasa-python/blob/42766d6fdae06896caac19bfa523f767d978cfc2/src/structure.pyx#L497)